be2-scala/src/main/scala/ch/epfl/pop/storage/DbActor.scala

```diff
@@ -384,6 +386,50 @@ final case class DbActor(
     heartbeatMap
   }
 
+  private def generateRumorKey(senderPk: PublicKey, rumorId: Int): String = {
+    s"${storage.RUMOR_KEY}${senderPk.base64Data.data}${Channel.DATA_SEPARATOR}$rumorId"
+  }
+
+  private def generateRumorDataKey(senderPk: PublicKey): String = {
+    s"${storage.RUMOR_DATA_KEY}${senderPk.base64Data.data}"
+  }
+
+  @throws[DbActorNAckException]
+  private def readRumorData(senderPk: PublicKey): RumorData = {
+    Try(storage.read(generateRumorDataKey(senderPk))) match {
+      case Success(Some(json)) => RumorData.buildFromJson(json)
+      case Success(None)       => RumorData(List.empty)
+      case Failure(ex)         => throw ex
+    }
+  }
+
+  @throws[DbActorNAckException]
+  private def readRumors(desiredRumors: Map[PublicKey, List[Int]]): Map[PublicKey, Map[Int, Rumor]] = {
```

Comment:
I am wondering, isn't the rumor id already in `Rumor` ? Do you need a map of map ? (if it makes your life easier down the line then sure, keep it that way) 
```suggestion
  private def readRumors(desiredRumors: Map[PublicKey, List[Int]]): Map[PublicKey, List[Rumor]] = {
```

Reply:
My idea was that with this type we have an easy access of which rumors_ids have been recovered from memory, without the need to parse each entry to know the corresponding rumor_id. This is oriented towards what I expect readRumors to be used for. I expect that when we query for the rumors, having their rumor_id easily accessible would be useful. Do you think it is better to have a more "intuitive" readRumors, that returns a list of rumors, and then the additionnal work would be done by the receivers ? Or to already have this type that already makes a small step of what we expect would be done after ?

Reply:
Following "You aren't gonna need it", I would say keep it as simple as possible unless you already know it will be useful.

If you think you will often need to check for the the presence of an ID then a Map could make sense, it will avoid the need of building the ID set at the cost of some small amount of memory and a more complex function signature.

I am fine with either, I won't block the pr for that and it can always be changed later anyway ;)

Reply:
I will follow your advice and go for a more intuitive function signature.